contracts/fCash/WrappedfCash.sol

```diff
@@ -0,0 +1,518 @@
+        require(msg.sender == address(NotionalV2), "Invalid caller");
+        // Only accept the fcash id that corresponds to the listed currency and maturity
+        uint256 fCashID = getfCashId();
+        require(_id == fCashID, "Invalid fCash asset");
+        // Protect against signed value underflows
+        require(int256(_value) > 0, "Invalid value");
+
+        // Double check the account's position, these are not strictly necessary and add gas costs
+        // but might be good safe guards
+        AccountContext memory ac = NotionalV2.getAccountContext(address(this));
+        require(ac.hasDebt == 0x00, "Incurred debt");
+        PortfolioAsset[] memory assets = NotionalV2.getAccountPortfolio(
+            address(this)
+        );
+        require(assets.length == 1, "Invalid assets");
+        require(
+            EncodeDecode.encodeERC1155Id(
+                assets[0].currencyId,
+                assets[0].maturity,
+                assets[0].assetType
+            ) == fCashID,
+            "Invalid portfolio asset"
+        );
+
+        // Update per account fCash balance, calldata from the ERC1155 call is
+        // passed via the ERC777 interface.
+        bytes memory userData;
+        bytes memory operatorData;
+        if (_operator == _from) userData = _data;
+        else operatorData = _data;
+
+        // We don't require a recipient ack here to maintain compatibility
+        // with contracts that don't support ERC777
+        _mint(_from, _value, userData, operatorData, false);
+
+        // This will allow the fCash to be accepted
+        return ERC1155_ACCEPTED;
+    }
+
+    /// @dev Do not accept batches of fCash
+    function onERC1155BatchReceived(
+        address, /* _operator */
+        address, /* _from */
+        uint256[] calldata, /* _ids */
+        uint256[] calldata, /* _values */
+        bytes calldata /* _data */
+    ) external pure override returns (bytes4) {
+        return 0;
+    }
+
+    /***** Redeem (Burn) Methods *****/
+
+    function redeem(uint256 amount, RedeemOpts memory opts) public override {
+        bytes memory data = abi.encode(opts);
+        burn(amount, data);
+    }
+
+    function redeemToAsset(uint256 amount, address receiver) external override {
+        redeem(amount, RedeemOpts({
+            redeemToUnderlying: false,
+            transferfCash: false,
+            receiver: receiver
+        }));
+    }
+
+    function redeemToUnderlying(uint256 amount, address receiver) external override {
+        redeem(amount, RedeemOpts({
+            redeemToUnderlying: true,
+            transferfCash: false,
+            receiver: receiver
+        }));
+    }
+
+    /// @notice Called before tokens are burned (redemption) and so we will handle
+    /// the fCash properly before and after maturity.
+    function _burn(
+        address from,
+        uint256 amount,
+        bytes memory userData,
+        bytes memory operatorData
```

Comment:
Can we add a comment here explaining how `userData` and `operatorData` are meant to be used?